webhook/rollout/validating/rollout_create_update_handler.go

```diff
@@ -0,0 +1,279 @@
+	errList = append(errList, validateRolloutSpecStrategy(&rollout.Spec.Strategy, fldPath.Child("Strategy"))...)
+	return errList
+}
+
+func validateRolloutSpecObjectRef(objectRef *appsv1alpha1.ObjectRef, fldPath *field.Path) field.ErrorList {
+	switch objectRef.Type {
+	case "", appsv1alpha1.WorkloadRefType:
+		if objectRef.WorkloadRef == nil ||
+			objectRef.WorkloadRef.Kind != "Deployment" ||
+			objectRef.WorkloadRef.APIVersion != apps.SchemeGroupVersion.String() {
+			return field.ErrorList{field.Invalid(fldPath.Child("WorkloadRef"), objectRef.WorkloadRef, "WorkloadRef only support 'Deployments.apps/v1'")}
+		}
+	default:
+		return field.ErrorList{field.Invalid(fldPath.Child("Type"), objectRef.Type, "ObjectRef only support 'workloadRef' type")}
+	}
+	return nil
+}
+
+func validateRolloutSpecStrategy(strategy *appsv1alpha1.RolloutStrategy, fldPath *field.Path) field.ErrorList {
+	switch strategy.Type {
+	case "", appsv1alpha1.RolloutStrategyCanary:
+		return validateRolloutSpecCanaryStrategy(strategy.Canary, fldPath.Child("Canary"))
+	default:
+		return field.ErrorList{field.Invalid(fldPath.Child("Type"), strategy.Type, "Strategy type only support 'canary'")}
+	}
+}
+
+func validateRolloutSpecCanaryStrategy(canary *appsv1alpha1.CanaryStrategy, fldPath *field.Path) field.ErrorList {
+	if canary == nil {
+		return field.ErrorList{field.Invalid(fldPath, nil, "Canary cannot be empty")}
+	}
+
+	errList := validateRolloutSpecCanarySteps(canary.Steps, fldPath.Child("Steps"))
+	errList = append(errList, validateRolloutSpecCanaryTraffic(canary.TrafficRouting, fldPath.Child("TrafficRouting"))...)
+	return errList
+}
+
+func validateRolloutSpecCanaryTraffic(traffic *appsv1alpha1.TrafficRouting, fldPath *field.Path) field.ErrorList {
+	if traffic == nil {
+		return field.ErrorList{field.Invalid(fldPath, nil, "Canary.TrafficRouting cannot be empty")}
+	}
+
+	errList := field.ErrorList{}
+	if len(traffic.Service) == 0 {
+		errList = append(errList, field.Invalid(fldPath.Child("Service"), traffic.Service, "TrafficRouting.Service cannot be empty"))
+	}
+
+	switch traffic.Type {
+	case "", appsv1alpha1.TrafficRoutingNginx:
+		if traffic.Nginx == nil ||
+			len(traffic.Nginx.Ingress) == 0 {
+			errList = append(errList, field.Invalid(fldPath.Child("Nginx"), traffic.Nginx, "TrafficRouting.Nginx.Ingress cannot be empty"))
+		}
+	default:
+
+		errList = append(errList, field.Invalid(fldPath.Child("Type"), traffic.Type, "TrafficRouting only support 'nginx' type"))
+	}
+
+	return errList
+}
+
+func validateRolloutSpecCanarySteps(steps []appsv1alpha1.CanaryStep, fldPath *field.Path) field.ErrorList {
+	stepCount := len(steps)
+	if stepCount == 0 {
+		return field.ErrorList{field.Invalid(fldPath, steps, "The number of Canary.Steps cannot be empty")}
+	}
+
+	if steps[stepCount-1].Weight != 100 {
+		return field.ErrorList{field.Invalid(fldPath, steps, "The 'Weight' field of the last CanaryStep must be '100'")}
```

Comment:
it is not necessary, and in most cases, actually the best practice is to keep a small fraction of canary application instances

Reply:
yes, this restrict is temporary. Now we can remove it.